### DISCUSSION TITLE
How to update global settings from plugin main.js?

### DISCUSSION THREAD

**Discussion by Muhomorik** [elgatosf/elgato#101] [Help] [answered]
Hi,

v. 6.5.1 (19929)

I am upgrading existing plugin written in plain JS and need to update my global setting  to the new structure.
It would be nice to have it done only once, when the plugin is updated or SD App restarted.

My idea was to do it in didReceiveGlobalSettings, update data and save settings again.
Receive didReceiveGlobalSettings and use correct object.

All PI will get the right object at registration.

`if (event === 'didReceiveGlobalSettings') {
	console.log("Main, didReceiveGlobalSettings.")

	// Get global settings
	const tempGlobalSettings = getSdGlobalSettings(jsonPayload['settings']);

	if (tempGlobalSettings.MigrationRequired()) {
		console.warn('Global settings migration is required.')

		tempGlobalSettings.Migrate()

		setTimeout(function(){
			console.log("Timer 2000");

			saveGlobalSettingsObject2(inPluginUUID, tempGlobalSettings)
			//saveGlobalSettings(inPluginUUID)

			console.warn('Global settings migration done.')

	}, 2000);
}`

However, I am not receiving _didReceiveGlobalSettings_ after saving it. I have even tried to use timeout to delay saving.

I can save global settings from PI, but not from plugin.

Are there any limitations or better way to do it?

**Comment by ChekTek** [comment]
Hi @Muhomorik,

Are you using the template repository (with an HTML file) or the new Node.js beta? 

Using the new Node.js beta, you can set global settings from the plugins like this:
```javascript
import streamDeck from "@elgato/streamdeck";

streamDeck.settings.setGlobalSettings({ myCoolSetting: "Hello World"});
```

if you are still using the old javascript library running in the Qt webengine, you should be able to set global settings like this:
```javascript
$SD.setGlobalSettings({ myCoolSetting: "Hello World"});
```

**Comment by Muhomorik** [comment]
Hi,
I am actually on pure js code that was made before SDK was made, but I can see something like that with sdk as well.


    const myAction = new Action('com.iccms.dmitri-1.action-test-1');

    let isSet = false;

    /**
     * The first event fired when Stream Deck starts
     */
    $SD.onConnected(({actionInfo, appInfo, connection, messageType, port, uuid}) => {
        console.log('Stream Deck connected!');

        $SD.getGlobalSettings();
    });

    /**
     * The first event fired when Stream Deck starts
     */
    $SD.onDidReceiveGlobalSettings(({actionInfo, appInfo, connection, messageType, port, uuid}) => {
        console.log(`Stream Deck onDidReceiveGlobalSettings. Set: ${isSet}.`);

        if (!isSet) {
            console.log("Update global settings.")
            isSet = true;

            $SD.setGlobalSettings({
                testId: 'test 1',
                testName: ' test name'
            })
        }
    });

    myAction.onKeyUp(({action, context, device, event, payload}) => {
        console.log('Your key code goes here!');
    });

    myAction.onDialRotate(({action, context, device, event, payload}) => {
        console.log('Your dial code goes here!');
    });


And the output I am getting in console is

![image](https://github.com/elgatosf/elgato/assets/11274409/36bf75e8-d3c3-4564-99af-4aa08450cb8b)

**Reply by ChekTek** [reply, marked as answer]
It looks like you using `setGlobalSettings` in your `didReceiveSettings` callback, but that callback is only triggered when `getGlobalSettings` is called. You'll want to set your settings before you call `getGlobalSettings` in order to see them in the event. 
```
$SD.onConnected((jsn) => {
	$SD.setGlobalSettings({ global: 'global' });
	$SD.getGlobalSettings();
});

$SD.onDidReceiveGlobalSettings((jsn) => {
	console.log(jsn);
});
```
